### ISSUE TITLE
Bug with simple_query_string, minimum_should_match, and multiple fields.

### ISSUE THREAD

**Issue by davidlbowen** [elastic/elasticsearch#13884]
```
# With a one-word query and minimum_should_match=-50%, adding extra non-matching fields should not matter.
# Tested on v1.7.2.

# delete and re-create the index
curl -XDELETE localhost:9200/test
curl -XPUT localhost:9200/test

echo 

 # insert a document
curl -XPUT 'http://localhost:9200/test/test/1' -d '
 { "title": "test document"}
 '
curl -XPOST 'http://localhost:9200/test/_refresh'

echo 

 # this correctly finds the document (f1 is a non-existent field)
 curl -XGET 'http://localhost:9200/test/test/_search' -d '{
  "query" : {
    "simple_query_string" : {
      "fields" : [ "title", "f1" ],
      "query" : "test",
      "minimum_should_match" : "-50%"
    }
  }
 }
'

echo 

# this incorrectly does not find the document (f1 and f2 are non-existent fields)
curl -XGET 'http://localhost:9200/test/test/_search' -d '{
  "query" : {
    "simple_query_string" : {
      "fields" : [ "title", "f1", "f2" ],
      "query" : "test",
      "minimum_should_match" : "-50%"
    }
  }
 }
'

echo
```

**Comment by clintongormley** [comment]
This does look like a bug.  The min_should_match is being applied at the wrong level:

    GET /test/test/_validate/query?explain
    {
      "query" : {
        "simple_query_string" : {
          "fields" : [ "title", "f1", "f2" ],
          "query" : "test",
          "minimum_should_match" : "-50%"
        }
      }
    }

Returns an explanation of:

    +((f1:test title:test f2:test)~2)

While the `query_string` and `multi_match` equivalents return:

    +(title:test | f1:test | f2:test)

**Comment by cbuescher** [comment]
I had a look and saw that `simple_query_string` iterates over all fields for each token in the query string and combines them all in boolean query `should` clauses, and we apply the `minimum_should_match` on the whole result. 

@clintongormley As far as I understand you, we should parse parse the query string for each field separately, apply the `minimum_should_match` there and then combine the result in an overall Boolean query. This however raised another question for me. Suppose we have two terms like `"query" : "test document"` instead, then currently we we get:

'''
(f1:test title:test f2:test) (f1:document title:document f2:document))~1
'''

If we would instead create the query per field individually we would get something like

'''
((title:test title:document)~1 (f1:test f1:document)~1 (f2:test f2:document)~1)
'''

While treating the query string for each field indicidually looks like the right behaviour in this case, I wonder if this will break other cases. wdyt?

**Comment by clintongormley** [comment]
@cbuescher the `query_string` query takes the same approach as your first output, ie:

    ((f1:test title:test f2:test) (f1:document title:document f2:document))~1

I think the bug is maybe a bit more subtle.  A query across 3 fields for two terms with min should match 80% results in:

    bool:
        min_should_match: 80% (==1)
        should:
          bool:
            should: [ f1:term1, f2:term1, f3:term1]
          bool:
            should: [ f1:term2, f2:term2, f3:term2]

however with only one term it is producing:

    bool:
      min_should_match: 80% (==2) 
      should: [ f1:term1, f2:term1, f3:term1]

In other words, min should match is being applied to the wrong `bool` query.  Instead, even the one term case should be wrapped in another `bool` query, and the min should match should be applied at that level.

**Comment by cbuescher** [comment]
@clintongormley Yes, I think thats what I meant. I'm working on a PR that applies the `minimum_should_match` to sub-queries that only target one field. That way your examples above would change to something like

```
bool:    
    should:
      bool:
        min_should_match: 80% (==1)
        should: [ f1:term1, f1:term2]
      bool:
        min_should_match: 80% (==1)
        should: [ f2:term1, f2:term2]
      bool:
        min_should_match: 80% (==1)
        should: [ f3:term1, f3:term2]
```
and for one term

```
bool:    
    should:
      bool:
        min_should_match: 80% (==0)
        should: [ f1:term1]
      bool:
        min_should_match: 80% (==0)
        should: [ f2:term1]
      bool:
        min_should_match: 80% (==0)
        should: [ f3:term1]
```
In the later case we already additionally simplify one-term bool queries to TermQueries.

**Comment by clintongormley** [comment]
@cbuescher I think that is incorrect.  The simple query string query (like the query string query) is term-centric rather than field-centric.  In other words, min should match should be applied to the number of terms (regardless of which field the term is in).

I'm guessing that there is an "optimization" for the one term case where the field-level bool clause is not wrapped in an outer bool clause.  Then the min should match is applied at the field level instead of at the term level, resulting in the wrong calculation.

**Comment by cbuescher** [comment]
That guess seems right, there is an optimization in lucenes
SimpleQueryParser for boolean queries with 0 or 1 clauses that seems to be
the problem. I think we can overwrite that.

On Thu, Oct 15, 2015 at 1:23 PM, Clinton Gormley <notifications@github.com>
wrote:

> @cbuescher <https://github.com/cbuescher> I think that is incorrect. The
> simple query string query (like the query string query) is term-centric
> rather than field-centric. In other words, min should match should be
> applied to the number of terms (regardless of which field the term is in).
>
> I'm guessing that there is an "optimization" for the one term case where
> the field-level bool clause is not wrapped in an outer bool clause. Then
> the min should match is applied at the field level instead of at the term
> level, resulting in the wrong calculation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/issues/13884#issuecomment-148358711>
> .
>



-- 
Christoph Büscher

**Comment by clintongormley** [comment]
If this is in Lucene, perhaps it should be fixed there?

@jdconrad what do you think?

**Comment by cbuescher** [comment]
@clintongormley I don't think SimpleQueryParser#simplify() is at the root of this anymore. The problem seems to be that SimpleQueryParser parses term by term-centric, but only starts wrapping the resulting queries when combining more than two of them. For one search term and two fields I get a Boolean query with two TermQuery clauses (without enclosing Boolean query), for two terms and one field I get an enclosing Boolean query with two Boolean query subclauses. I'm not sure yet how this can be distiguished from outside of the Lucene parser without inspecting the query, and if a solution like that holds for more complicated cases.

**Comment by cbuescher** [comment]
Althought it would be nice if Lucene SimpleQueryParse would output a Boolquery with one should-clause and three nested Boolqueries for the 1-term/multi-field case, I think we can detect this case and do the wrapping in the additional Boolquery in the SimpleQueryStringBuilder. I just opened a PR.

**Comment by jdconrad** [comment]
The SQP wasn't really designed around multi-field terms, but needed to have it added afterwards for use as a default field which is why the min-should-match never gets applied down at that the level.  I don't know if the correct behavior is to make it work on multi-fields.  I'll have to give that some thought given that it really is as @cbuescher described as term-centric, and it sort of supposed to be disguised from the user.  One thing that will make this easier to fix, though, I believe is #4707, since it will flatten the parse tree a bit.

**Comment by cbuescher** [comment]
@jdconrad thanks for explaining, in the meantime I opened #14186 which basically tries to distinguish the one vs. multi-field cases and tries wraps the resulting query one more time to get a correct min-should-match. Please leave comment there if my current approach will colide the plans regarding #4707.

**Comment by cbuescher** [comment]
Reopening this issue since the fix proposed in #14186 was too fragile. Discussed with @javanna and @jpountz, at this point we think the options are either fixing this in lucenes SimpleQueryParser so that we can apply minimum_should_match correctly on the ES side or remove this option from `simple_query_string` entirely because it cannot properly supported.

**Comment by cbuescher** [comment]
Trying to sum up this issue so far: 

* the number of should-clauses returned by `SimpleQueryParser` is not 1 for one search term and multiple fields, so we cannot apply `minimum_should_match` correctly in `SimpleQueryStringBuilder`. e.g. for `"query" : "term1", "fields" : [ "f1", "f2", "f3" ]` SimpleQueryParser returns a BooleanQuery with three should-clauses. As soon  as we add more search terms, the number of should-clauses is the same as the number of search terms, e.g. `"query" : "term1 term2", "fields" : [ "f1", "f2", "f3" ]` returns a BooleanQuery with two subclauses, one per term.
* it is difficult to determine the number of terms from the query string upfront, because the tokenization depends on the analyzer used, so we really need `SimpleQueryParser#parse()` for this.
* it is hard to determine the correct number of terms from the returned lucene query without making assumptions about the inner structure of the query (which is subject to change, reason for #14186 beeing reverted). e.g. currently `"query" : "term1", "fields" : [ "f1", "f2", "f3" ]` and `"query" : "term1 term2 term3", "fields" : [ "f1" ]` will return a BooleanQuery with same structure (three should-clauses, each containing a TermQuery).

**Comment by jimferenczi** [comment]
@cbuescher this issue is fixed by https://github.com/elastic/elasticsearch/pull/16155. 
@rmuir has pointed out a nice way to distinguish between a single word query with multiple fields against a multi word query with a single field: we just have to check if the coord are disabled on the top level BooleanQuery, the simple query parser disables the coord when the boolean query for multiple fields is built.
Though I tested the single word with multiple fields case only, if you think of other issues please reopen this ticket or open a new one ;).

**Comment by cbuescher** [comment]
@jimferenczi thats great, I just checked this with a test which is close to the problem desciption here. I'm not sure if this adds anything to your tests, but just in case I justed opened  #16465 which adds this as an integration test for SimpleQueryStringQuery. Maybe you can take a look and tell me if it makes sense to add those as well.

**Comment by jimferenczi** [comment]
@cbuescher thanks, a unit test in SimpleQueryStringBuilderTest could be useful as well. The integ test does not check the minimum should match that is applied (or not) to the boolean query.
